File: Algorithms/p106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/p106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        unordered_map<int,int> index;
        for(int i=0;i<inorder.size();++i) index[inorder[i]] = i;
        return buildTreeHelper(inorder,0,inorder.size(),postorder,0,postorder.size(),index);
    }
    TreeNode* buildTreeHelper(vector<int>& inorder, int l1, int r1, vector<int>& postorder, int l2, int r2, unordered_map<int,int>& index) {
        if(l1>=r1) return 0;
        int i = index[postorder[r2-1]];
        TreeNode* root = new TreeNode(inorder[i]);
        root->left = buildTreeHelper(inorder,l1,i,postorder,l2,l2+i-l1,index);
        root->right = buildTreeHelper(inorder,i+1,r1,postorder,l2+i-l1,r2-1,index);
        return root;
    }
};

File: Algorithms/p106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal/p106_Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        index = {}
        for i in range(len(inorder)):
            index[inorder[i]] = i
        return self.buildTreeHelper(inorder,0,len(inorder),postorder,0,len(postorder),index)

    def buildTreeHelper(self, inorder, l1, r1, postorder, l2, r2, index):
        if l1>=r1:
            return None
        i = index[postorder[r2-1]]
        root = TreeNode(inorder[i])
        root.left = self.buildTreeHelper(inorder,l1,i,postorder,l2,l2+i-l1,index)
        root.right = self.buildTreeHelper(inorder,i+1,r1,postorder,l2+i-l1,r2-1,index)
        return root
